src/server/transport/socketio.ts

```diff
@@ -126,7 +130,12 @@ export class SocketIO {
             TransportAPI(matchID, socket, this.clientInfo, this.roomInfo),
             this.auth
           );
-          await master.onUpdate(action, stateID, matchID, playerID);
+
+          const queueInterval = IsSynchronous(app.context.db) ? 300 : 0;
```

Comment:
I did that because the problem continued to occur with InMemory, even using queue. That interval solved it. Maybe a smaller interval should work, have to test. I'll run more tests for this point.

Reply:
> This is complicated because the unit tests will not always reflect the real world.

Yes, that’s my fear — anything interval-based is dependent on the real world staying within that interval, which it might not. It also suggests there’s a more fundamental bug in the queueing mechanism. If the queue is not waiting correctly to process each update with synchronous storage (which shouldn’t happen concurrently anyway 😕?), adding an interval “fixes” it kind of, but I’m guessing only where the update processes fast enough, so we still have a race condition on our hands if the master ever slows down enough.

We probably need to understand why the `InMemory` storage breaks the queuing system, rather than patching it with an interval.

Reply:
I understand your point. I believe that writing in memory happens too fast and it does not have a "await" functionality, so an action's result overwrite another action result because they seem to happen at the same time in memory. Right now, I think that adding logs to InMemory will help to investigate that.

Reply:
Yes, some logging to check which steps are happening before/after each other without the interval could be a good idea. Some strange order things can happen with mixed async and sync stuff like:

```js
const f1 = async () => {
  await somethingThatTakesTime();
  console.log('f1');
};

const f2 = () => {
  f1(); // not awaited
  console.log('f2');
};

f2();
// output:
// 'f2'
// 'f1'
```

In a simple example like this it might be obvious, but it’s easy to get tangled up in this with a more complex codebase like this one.

Reply:
Even if you await for 0 seconds on f1, it will be printed out after f2.  :confused:

Reply:
Yup, exactly. That’s expected, but does need careful management.